enigma-runtime-t/src/lib.rs

```diff
@@ -185,6 +205,36 @@ impl Runtime {
         Ok(())
     }
 
+    /// args:
+   /// * `new_value_len` - the length of of the new value to be written under the `key`
+   /// * `key` - the key to write the new value
+   /// calculate the gas to be charged for the writing of the new value.
+   /// There is an initial constant value charged for the writing
+   /// If the new value is larger than the old one, then gas is charged for each new byte.
+   /// If the new value is smaller than the old one, then the charge for each saved byte
+   /// is subtracted from the initial value until 0.
+    fn calculate_gas_for_writing(&mut self, new_value_len: u32, key: &str) -> Result<u64> {
+        let mut result = 0;
+        let val = &self.post_execution_state.json[key];
+        let mut old_value_len = 0;
+        // forcing the length of Null value to be 0, since it is not 0.
+        if !val.is_null() {
+            let old_value_vec =
+                serde_json::to_vec(val).expect("Failed converting Value to vec in Runtime while reading state");
+            old_value_len = old_value_vec.len() as u32;
+        }
+        if new_value_len >= old_value_len {
+            result = self.gas_costs.write_value + (new_value_len - old_value_len)*self.gas_costs.write_additional_byte;
+        }
+        else {
+            let decrease_cost = (old_value_len - new_value_len)*self.gas_costs.write_additional_byte;
+            if decrease_cost < self.gas_costs.write_value {
+                result = self.gas_costs.write_value - decrease_cost;
```

Comment:
what if the `decrease_cost` is higher than write_value? should this remove from the total amount of gas? or just this gas?

in general idk about this separation of gases, but maybe this is something that should be handle when we redo the runtime.

Reply:
It's a good idea to decrease from the global gas. I changed the code accordingly.